File: stdlib/js/hull/middleware/health.js
/**
 * @file hull:middleware:health
 * @module hull:middleware:health
 * @description Liveness + readiness HTTP endpoints. Lua parity:
 *   `hull.middleware.health`.
 *
 * `health.middleware()` intercepts `GET /health` and `GET /ready`:
 *
 *   - `/health` (liveness) — always returns `200 {status:"ok", uptime}`.
 *     Verifies the process is alive; never fails.
 *   - `/ready`  (readiness) — runs the DB ping (if enabled) and every
 *     check registered via `health.register()`. Returns `200` when all
 *     pass, otherwise `503` with a per-check breakdown.
 *
 * ES modules can't conditionally import — pass the `db` module
 * explicitly via `health.setDb(dbModule)` or `opts.db`.
 *
 * @license AGPL-3.0-or-later
 */

import { time } from "hull:time";
import { server } from "hull:server";

/* db module is optional — not all apps use a database. */
let _dbMod = null;

const _checks = {};
let _startTime = null;

/**
 * Register a custom readiness check.
 *
 * Replaces any prior check with the same name. Contract:
 *   - return `true` or no return value → ok
 *   - return `false` → fail
 *   - throw → fail (uses error string)
 *
 * @param {string}      name  Identifier (appears in `/ready` JSON output).
 * @param {() => boolean|void} fn
 */
function register(name, fn) {
    _checks[name] = fn;
}

/**
 * Remove a previously registered check.
 * @param {string} name
 */
function unregister(name) {
    delete _checks[name];
}

/**
 * Run the DB ping (if enabled) + every registered check.
 *
 * @param {Object} [opts]
 * @param {boolean} [opts.dbCheck=true]  Include the DB ping.
 * @returns {{ checks: Object<string,{status:string,latency_ms:number,error?:string}>,
 *             allOk: boolean }}
 */
function runChecks(opts) {
    const o = opts || {};
    const results = {};
    let allOk = true;

    // DB check (db module may not be available if app has no database)
    if (o.dbCheck !== false && _dbMod) {
        const t0 = time.clock();
        try {
            _dbMod.query("SELECT 1");
            const latency = Math.round((time.clock() - t0) * 10) / 10;
            results.db = { status: "ok", latency_ms: latency };
        } catch (e) {
            const latency = Math.round((time.clock() - t0) * 10) / 10;
            results.db = { status: "fail", error: String(e), latency_ms: latency };
            allOk = false;
        }
    }

    // Custom checks
    for (const name in _checks) {
        const t0 = time.clock();
        try {
            const result = _checks[name]();
            const latency = Math.round((time.clock() - t0) * 10) / 10;
            if (result === false) {
                results[name] = { status: "fail", error: "check returned false", latency_ms: latency };
                allOk = false;
            } else {
                results[name] = { status: "ok", latency_ms: latency };
            }
        } catch (e) {
            const latency = Math.round((time.clock() - t0) * 10) / 10;
            results[name] = { status: "fail", error: String(e), latency_ms: latency };
            allOk = false;
        }
    }

    return { checks: results, allOk };
}

/**
 * Inject the db module used for the readiness DB ping.
 *
 * Equivalent to passing `opts.db` to `middleware()`.
 *
 * @param {Object} dbModule  The `hull:db` module value.
 */
function setDb(dbModule) {
    _dbMod = dbModule;
}

/**
 * Build the `/health` + `/ready` middleware.
 *
 * Returns `1` (short-circuit) when the request hits one of the health
 * paths and `0` (continue) otherwise.
 *
 * @param {Object} [opts]
 * @param {string} [opts.pathHealth="/health"]
 * @param {string} [opts.pathReady="/ready"]
 * @param {boolean}[opts.dbCheck=true]  Include DB ping in `/ready`.
 * @param {Object} [opts.db]  Inject the db module (alternative to `setDb`).
 * @returns {(req, res) => number}
 */
function middleware(opts) {
    const o = opts || {};
    const pathHealth = o.pathHealth || "/health";
    const pathReady = o.pathReady || "/ready";
    const doDbCheck = o.dbCheck !== false;

    /* Accept db module via opts for apps that have a database */
    if (o.db) _dbMod = o.db;

    if (_startTime === null) {
        _startTime = time.now();
    }

    return function healthMiddleware(req, res) {
        if (req.method !== "GET" && req.method !== "HEAD") {
            return 0;
        }

        const uptime = Math.floor(time.now() - _startTime);

        if (req.path === pathHealth) {
            res.json({ status: "ok", uptime });
            return 1;
        }

        if (req.path === pathReady) {
            const result = runChecks({ dbCheck: doDbCheck });

            const body = {
                status: result.allOk ? "ok" : "fail",
                checks: result.checks,
                uptime,
            };

            if (server && server.stats) {
                body.stats = server.stats();
            }

            if (result.allOk) {
                res.json(body);
            } else {
                res.status(503).json(body);
            }
            return 1;
        }

        return 0;
    };
}

const health = { register, unregister, runChecks, middleware, setDb };
export { health };

File: stdlib/lua/hull/middleware/health.lua
--- Liveness + readiness HTTP endpoints.
--
-- `health.middleware()` intercepts `GET /health` and `GET /ready`:
--
--   - `/health` (liveness) — always returns `200 {status="ok", uptime}`.
--     Verifies the process is alive; never fails.
--   - `/ready`  (readiness) — runs the DB ping (if enabled) and every
--     check registered via `health.register()`. Returns `200` when all
--     pass, otherwise `503` with a per-check breakdown.
--
-- Custom check contract:
--   * `true` or no return value  → ok
--   * `false, "msg"`              → fail (optional message)
--   * raised error                → fail (uses error string)
--
-- @module hull.middleware.health
-- @license AGPL-3.0-or-later
-- @usage
--   local health = require("hull.middleware.health")
--   health.register("redis", function()
--       return redis.ping() == "PONG"
--   end)
--   app.use("GET", "/*", health.middleware())

local db = require("hull.db")
local server = require("hull.http-server")
local time = require("hull.time")

local health = {}

local _checks = {}
local _start_time = nil

--- Register a custom readiness check.
--
-- Replaces any prior check with the same name.
--
-- @function health.register
-- @tparam string name  Identifier (appears in `/ready` JSON output).
-- @tparam function fn  `function() -> ok[, msg]` — see contract above.
function health.register(name, fn)
    _checks[name] = fn
end

--- Remove a previously registered check.
-- @function health.unregister
-- @tparam string name
function health.unregister(name)
    _checks[name] = nil
end

--- Run the DB ping (if enabled) + every registered check.
--
-- @function health.run_checks
-- @tparam[opt] table opts
-- @tparam[opt=true] boolean opts.db_check  Include the DB ping.
-- @treturn table  `{ checks = { name = {status, latency_ms, error?}, ... },
--                  all_ok = boolean }`
function health.run_checks(opts)
    opts = opts or {}
    local results = {}
    local all_ok = true

    -- DB check
    if opts.db_check ~= false and db then
        local t0 = time.clock()
        local ok, err = pcall(db.query, "SELECT 1")
        local latency = math.floor((time.clock() - t0) * 10 + 0.5) / 10
        if ok then
            results.db = { status = "ok", latency_ms = latency }
        else
            results.db = { status = "fail", error = tostring(err), latency_ms = latency }
            all_ok = false
        end
    end

    -- Custom checks. Contract: fn() returns
    --   true               → ok
    --   (no return / nil)  → ok (back-compat — common Lua idiom for
    --                        success when an assert-style check passes)
    --   false[, "msg"]     → fail (optional message)
    --   true, "info"       → ok with extra info (ignored)
    --   other (number/etc) → fail to surface bugs
    --
    -- Phase 6 audit M-1: the original Phase 6 patch tightened the
    -- contract to "only `ret == true` is ok", which silently flipped
    -- every existing `function() end` check to fail. Operationally
    -- dangerous on upgrade (readiness probes go to 503). Restored
    -- back-compat: nil-on-success and bare-true are both ok.
    for name, fn in pairs(_checks) do
        local t0 = time.clock()
        local ok, ret, err_msg = pcall(fn)
        local latency = math.floor((time.clock() - t0) * 10 + 0.5) / 10
        if not ok then
            -- Check threw — `ret` is the error message.
            results[name] = { status = "fail",
                              error = type(ret) == "string" and ret or "check threw",
                              latency_ms = latency }
            all_ok = false
        elseif ret == nil or ret == true then
            -- Success: explicit `true` or implicit (no return value).
            results[name] = { status = "ok", latency_ms = latency }
        else
            -- ret is false, a string, a number, etc. — treat as failure.
            local msg = "check failed"
            if type(err_msg) == "string" then msg = err_msg
            elseif type(ret) == "string" then msg = ret end
            results[name] = { status = "fail", error = msg, latency_ms = latency }
            all_ok = false
        end
    end

    return { checks = results, all_ok = all_ok }
end

--- Build the `/health` + `/ready` middleware.
--
-- Returns `1` (short-circuit) when the request hits one of the health
-- paths and `0` (continue) otherwise. Records the start time on first
-- invocation so subsequent `/health` responses include uptime.
--
-- @function health.middleware
-- @tparam[opt] table opts
-- @tparam[opt="/health"] string opts.path_health  Liveness path.
-- @tparam[opt="/ready"]  string opts.path_ready   Readiness path.
-- @tparam[opt=true]      boolean opts.db_check    Include DB ping in `/ready`.
-- @treturn function(req, res) -> 0|1
function health.middleware(opts)
    opts = opts or {}
    local path_health = opts.path_health or "/health"
    local path_ready  = opts.path_ready or "/ready"
    local do_db_check = opts.db_check ~= false

    if not _start_time then
        _start_time = time.now()
    end

    return function(req, res)
        -- Only handle GET/HEAD
        if req.method ~= "GET" and req.method ~= "HEAD" then
            return 0
        end

        local uptime = math.floor(time.now() - _start_time)

        if req.path == path_health then
            res:json({ status = "ok", uptime = uptime })
            return 1
        end

        if req.path == path_ready then
            local result = health.run_checks({ db_check = do_db_check })

            local body = {
                status = result.all_ok and "ok" or "fail",
                checks = result.checks,
                uptime = uptime,
            }

            -- Include server stats if available
            if server and server.stats then
                body.stats = server.stats()
            end

            if result.all_ok then
                res:json(body)
            else
                res:status(503):json(body)
            end
            return 1
        end

        return 0
    end
end

return health
